Library/Homebrew/tap.rb

```diff
@@ -171,21 +171,42 @@ def core_tap?
   # @option options [Boolean] :quiet If set, suppress all output.
   def install(options = {})
     require "descriptions"
-    raise TapAlreadyTappedError, name if installed?
-    clear_cache
 
+    full_clone = options.fetch(:full_clone, true)
     quiet = options.fetch(:quiet, false)
 
+    if installed?
+      raise TapAlreadyTappedError, name unless full_clone
+      unless (path/".git/shallow").exist?
+        ohai "#{name} is already a full clone" unless quiet
+        return
+      end
+    end
+
+    clear_cache unless installed?
+
     # ensure git is installed
     Utils.ensure_git_installed!
-    ohai "Tapping #{name}" unless quiet
+    if installed?
+      ohai "Unshallowing #{name}" unless quiet
+    else
+      ohai "Tapping #{name}" unless quiet
+    end
     remote = options[:clone_target] || "https://github.com/#{user}/homebrew-#{repo}"
-    args = %W[clone #{remote} #{path} --config core.autocrlf=false]
-    args << "--depth=1" unless options.fetch(:full_clone, false)
+    args = if installed?
+      %W[fetch --unshallow]
+    else
+      %W[clone #{remote} #{path} --config core.autocrlf=false]
+    end
+    args << "--depth=1" unless full_clone
     args << "-q" if quiet
 
     begin
-      safe_system "git", *args
+      if installed?
+        path.cd { safe_system "git", *args }
```

Comment:
I think we should not delete the tap directory when something goes wrong. So you don't need to add it in the begin rescue block.

In fact, for unshallow should do the things like `link_manpages` or cache the formulae description in the below. I suggest we treat this case separately:
```ruby
def install(options = {})
  ...
  if installed? 
    raise TapAlreadyTappedError, name unless full_clone
    unless shallow?
      ...
    end
      ... put unshallow code here ...
    return
  end
  
  .. rest are the same as before ...
end
```